File: backend/src/controllers/timerController.ts
import { Request, Response } from 'express';
import { TimerService, TimerConfig } from '../services/timerService';
import { PrismaClient } from '@prisma/client';

const prisma = new PrismaClient();

/**
 * Controller for study session timer operations
 */
export class TimerController {
  private timerService: TimerService;

  constructor() {
    this.timerService = new TimerService();
  }

  /**
   * Start or resume timer for a session
   * POST /api/sessions/:questionSetId/timer/start
   */
  startTimer = async (req: Request, res: Response): Promise<void> => {
    try {
      if (!req.user) {
        res.status(401).json({ error: 'User not authenticated' });
        return;
      }

      const questionSetId = parseInt(req.params.questionSetId);
      const userId = req.user.userId;
      const config: Partial<TimerConfig> = req.body;

      console.log('🎯 TimerController: Starting timer for questionSet', questionSetId, 'user', userId);

      // Find the active session for this user and question set
      const sessionId = await this.getActiveSessionId(userId, questionSetId);

      const timerState = await this.timerService.startTimer(sessionId, config);

      res.json({
        success: true,
        timer: timerState
      });
    } catch (error) {
      console.error('Start timer error:', error);
      res.status(500).json({ error: 'Internal server error' });
    }
  };

  /**
   * Pause timer for a session
   * POST /api/sessions/:questionSetId/timer/pause
   */
  pauseTimer = async (req: Request, res: Response): Promise<void> => {
    try {
      if (!req.user) {
        res.status(401).json({ error: 'User not authenticated' });
        return;
      }

      const questionSetId = parseInt(req.params.questionSetId);
      const userId = req.user.userId;

      console.log('⏸️ TimerController: Pausing timer for questionSet', questionSetId);

      const sessionId = await this.getActiveSessionId(userId, questionSetId);
      const timerState = await this.timerService.pauseTimer(sessionId);

      res.json({
        success: true,
        timer: timerState
      });
    } catch (error) {
      console.error('Pause timer error:', error);
      res.status(500).json({ error: 'Internal server error' });
    }
  };

  /**
   * Advance to next phase (work -> rest -> work)
   * POST /api/sessions/:questionSetId/timer/advance
   */
  advancePhase = async (req: Request, res: Response): Promise<void> => {
    try {
      if (!req.user) {
        res.status(401).json({ error: 'User not authenticated' });
        return;
      }

      const questionSetId = parseInt(req.params.questionSetId);
      const userId = req.user.userId;

      console.log('➡️ TimerController: Advancing phase for questionSet', questionSetId);

      const sessionId = await this.getActiveSessionId(userId, questionSetId);
      const timerState = await this.timerService.advancePhase(sessionId);

      res.json({
        success: true,
        timer: timerState
      });
    } catch (error) {
      console.error('Advance phase error:', error);
      res.status(500).json({ error: 'Internal server error' });
    }
  };

  /**
   * Stop timer and complete session
   * POST /api/sessions/:questionSetId/timer/stop
   */
  stopTimer = async (req: Request, res: Response): Promise<void> => {
    try {
      if (!req.user) {
        res.status(401).json({ error: 'User not authenticated' });
        return;
      }

      const questionSetId = parseInt(req.params.questionSetId);
      const userId = req.user.userId;

      console.log('⏹️ TimerController: Stopping timer for questionSet', questionSetId);

      const sessionId = await this.getActiveSessionId(userId, questionSetId);
      const timerState = await this.timerService.stopTimer(sessionId);

      res.json({
        success: true,
        timer: timerState
      });
    } catch (error) {
      console.error('Stop timer error:', error);
      res.status(500).json({ error: 'Internal server error' });
    }
  };

  /**
   * Get current timer state
   * GET /api/sessions/:questionSetId/timer
   */
  getTimerState = async (req: Request, res: Response): Promise<void> => {
    try {
      if (!req.user) {
        res.status(401).json({ error: 'User not authenticated' });
        return;
      }

      const questionSetId = parseInt(req.params.questionSetId);
      const userId = req.user.userId;

      const sessionId = await this.getActiveSessionId(userId, questionSetId);
      const timerState = await this.timerService.getTimerState(sessionId);

      res.json({
        success: true,
        timer: timerState
      });
    } catch (error) {
      console.error('Get timer state error:', error);
      res.status(500).json({ error: 'Internal server error' });
    }
  };

  /**
   * Get timer statistics for a session
   * GET /api/sessions/:questionSetId/timer/stats
   */
  getTimerStats = async (req: Request, res: Response): Promise<void> => {
    try {
      if (!req.user) {
        res.status(401).json({ error: 'User not authenticated' });
        return;
      }

      const questionSetId = parseInt(req.params.questionSetId);
      const userId = req.user.userId;

      const sessionId = await this.getActiveSessionId(userId, questionSetId);
      const stats = await this.timerService.getTimerStats(sessionId);

      res.json({
        success: true,
        stats
      });
    } catch (error) {
      console.error('Get timer stats error:', error);
      res.status(500).json({ error: 'Internal server error' });
    }
  };

  /**
   * Update timer configuration
   * PUT /api/sessions/:questionSetId/timer/config
   */
  updateConfig = async (req: Request, res: Response): Promise<void> => {
    try {
      if (!req.user) {
        res.status(401).json({ error: 'User not authenticated' });
        return;
      }

      const questionSetId = parseInt(req.params.questionSetId);
      const userId = req.user.userId;
      const config: Partial<TimerConfig> = req.body;

      console.log('⚙️ TimerController: Updating config for questionSet', questionSetId, 'with:', config);

      const sessionId = await this.getActiveSessionId(userId, questionSetId);
      const timerState = await this.timerService.updateConfig(sessionId, config);

      res.json({
        success: true,
        timer: timerState
      });
    } catch (error) {
      console.error('Update timer config error:', error);
      res.status(500).json({ error: 'Internal server error' });
    }
  };

  /**
   * Helper method to get active session ID for a user and question set
   * Creates a new session if none exists
   */
  private async getActiveSessionId(userId: number, questionSetId: number): Promise<number> {
    // First try to find an existing active session
    let session = await prisma.studySession.findFirst({
      where: {
        userId,
        questionSetId,
        completedAt: null,
      },
      select: { id: true }
    });

    // If no active session exists, create a new one
    if (!session) {
      console.log('⚡ TimerController: Creating new session for questionSet', questionSetId, 'user', userId);

      const newSession = await prisma.studySession.create({
        data: {
          userId,
          questionSetId,
          mode: 'front-to-end', // Default mode for timer-initiated sessions
        },
        select: { id: true }
      });

      session = newSession;
    }

    return session.id;
  }
}

File: backend-springboot/src/main/java/com/learningadvisor/backend/controller/TimerController.java
package com.learningadvisor.backend.controller;

import com.learningadvisor.backend.dto.timer.TimerConfigRequest;
import com.learningadvisor.backend.dto.timer.TimerStateResponse;
import com.learningadvisor.backend.dto.timer.TimerStatsResponse;
import com.learningadvisor.backend.entity.TimerSession;
import com.learningadvisor.backend.service.TimerService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.web.bind.annotation.*;

import java.util.HashMap;
import java.util.Map;

/**
 * REST controller for timer operations
 * Provides endpoints for Pomodoro-style study timers
 */
@RestController
@RequestMapping("/api/study-sessions/{questionSetId}/timer")
@RequiredArgsConstructor
@Slf4j
public class TimerController {

    private final TimerService timerService;

    /**
     * Start or resume timer for a study session
     * POST /api/study-sessions/{questionSetId}/timer/start
     */
    @PostMapping("/start")
    public ResponseEntity<Map<String, Object>> startTimer(
            @PathVariable Long questionSetId,
            @RequestBody(required = false) TimerConfigRequest config,
            @AuthenticationPrincipal UserDetails userDetails) {

        try {
            // Get user ID from username
            Long userId = Long.parseLong(userDetails.getUsername());

            log.info("Starting timer for user {} questionSet {}", userId, questionSetId);

            // Extract config values
            Integer workDuration = config != null ? config.getWorkDuration() : null;
            Integer restDuration = config != null ? config.getRestDuration() : null;
            Boolean isInfinite = config != null ? config.getIsInfinite() : null;

            TimerSession timerSession = timerService.startTimer(userId, questionSetId, workDuration, restDuration, isInfinite);
            TimerStateResponse response = TimerStateResponse.fromEntity(timerSession);

            Map<String, Object> result = new HashMap<>();
            result.put("timer", response);

            return ResponseEntity.ok(result);

        } catch (Exception e) {
            log.error("Failed to start timer", e);
            Map<String, Object> error = new HashMap<>();
            error.put("error", e.getMessage());
            return ResponseEntity.badRequest().body(error);
        }
    }

    /**
     * Pause timer for a study session
     * POST /api/study-sessions/{questionSetId}/timer/pause
     */
    @PostMapping("/pause")
    public ResponseEntity<Map<String, Object>> pauseTimer(
            @PathVariable Long questionSetId,
            @AuthenticationPrincipal UserDetails userDetails) {

        try {
            Long userId = Long.parseLong(userDetails.getUsername());

            log.info("Pausing timer for user {} questionSet {}", userId, questionSetId);

            TimerSession timerSession = timerService.pauseTimer(userId, questionSetId);
            TimerStateResponse response = TimerStateResponse.fromEntity(timerSession);

            Map<String, Object> result = new HashMap<>();
            result.put("timer", response);

            return ResponseEntity.ok(result);

        } catch (Exception e) {
            log.error("Failed to pause timer", e);
            Map<String, Object> error = new HashMap<>();
            error.put("error", e.getMessage());
            return ResponseEntity.badRequest().body(error);
        }
    }

    /**
     * Advance to next phase (work -> rest -> work)
     * POST /api/study-sessions/{questionSetId}/timer/advance
     */
    @PostMapping("/advance")
    public ResponseEntity<Map<String, Object>> advancePhase(
            @PathVariable Long questionSetId,
            @AuthenticationPrincipal UserDetails userDetails) {

        try {
            Long userId = Long.parseLong(userDetails.getUsername());

            log.info("Advancing timer phase for user {} questionSet {}", userId, questionSetId);

            TimerSession timerSession = timerService.advancePhase(userId, questionSetId);
            TimerStateResponse response = TimerStateResponse.fromEntity(timerSession);

            Map<String, Object> result = new HashMap<>();
            result.put("timer", response);

            return ResponseEntity.ok(result);

        } catch (Exception e) {
            log.error("Failed to advance timer phase", e);
            Map<String, Object> error = new HashMap<>();
            error.put("error", e.getMessage());
            return ResponseEntity.badRequest().body(error);
        }
    }

    /**
     * Stop timer and complete session
     * POST /api/study-sessions/{questionSetId}/timer/stop
     */
    @PostMapping("/stop")
    public ResponseEntity<Map<String, Object>> stopTimer(
            @PathVariable Long questionSetId,
            @AuthenticationPrincipal UserDetails userDetails) {

        try {
            Long userId = Long.parseLong(userDetails.getUsername());

            log.info("Stopping timer for user {} questionSet {}", userId, questionSetId);

            TimerSession timerSession = timerService.stopTimer(userId, questionSetId);
            TimerStateResponse response = TimerStateResponse.fromEntity(timerSession);

            Map<String, Object> result = new HashMap<>();
            result.put("timer", response);

            return ResponseEntity.ok(result);

        } catch (Exception e) {
            log.error("Failed to stop timer", e);
            Map<String, Object> error = new HashMap<>();
            error.put("error", e.getMessage());
            return ResponseEntity.badRequest().body(error);
        }
    }

    /**
     * Get current timer state
     * GET /api/study-sessions/{questionSetId}/timer
     */
    @GetMapping
    public ResponseEntity<Map<String, Object>> getTimerState(
            @PathVariable Long questionSetId,
            @AuthenticationPrincipal UserDetails userDetails) {

        try {
            Long userId = Long.parseLong(userDetails.getUsername());

            log.info("Getting timer state for user {} questionSet {}", userId, questionSetId);

            TimerSession timerSession = timerService.getTimerState(userId, questionSetId);
            TimerStateResponse response = TimerStateResponse.fromEntity(timerSession);

            Map<String, Object> result = new HashMap<>();
            result.put("timer", response);

            return ResponseEntity.ok(result);

        } catch (Exception e) {
            log.error("Failed to get timer state", e);
            Map<String, Object> error = new HashMap<>();
            error.put("error", e.getMessage());
            return ResponseEntity.badRequest().body(error);
        }
    }

    /**
     * Update timer configuration
     * PUT /api/study-sessions/{questionSetId}/timer/config
     */
    @PutMapping("/config")
    public ResponseEntity<Map<String, Object>> updateConfig(
            @PathVariable Long questionSetId,
            @RequestBody TimerConfigRequest config,
            @AuthenticationPrincipal UserDetails userDetails) {

        try {
            Long userId = Long.parseLong(userDetails.getUsername());

            log.info("Updating timer config for user {} questionSet {}", userId, questionSetId);

            TimerSession timerSession = timerService.updateConfig(
                    userId,
                    questionSetId,
                    config.getWorkDuration(),
                    config.getRestDuration(),
                    config.getIsInfinite()
            );

            TimerStateResponse response = TimerStateResponse.fromEntity(timerSession);

            Map<String, Object> result = new HashMap<>();
            result.put("timer", response);

            return ResponseEntity.ok(result);

        } catch (Exception e) {
            log.error("Failed to update timer config", e);
            Map<String, Object> error = new HashMap<>();
            error.put("error", e.getMessage());
            return ResponseEntity.badRequest().body(error);
        }
    }

    /**
     * Get timer statistics including event history
     * GET /api/study-sessions/{questionSetId}/timer/stats
     */
    @GetMapping("/stats")
    public ResponseEntity<Map<String, Object>> getTimerStats(
            @PathVariable Long questionSetId,
            @AuthenticationPrincipal UserDetails userDetails) {

        try {
            Long userId = Long.parseLong(userDetails.getUsername());

            log.info("Getting timer stats for user {} questionSet {}", userId, questionSetId);

            TimerService.TimerSessionStats stats = timerService.getTimerStats(userId, questionSetId);

            TimerStatsResponse response = TimerStatsResponse.builder()
                    .totalWorkTime(stats.getTotalWorkTime())
                    .totalRestTime(stats.getTotalRestTime())
                    .totalTime(stats.getTotalTime())
                    .cyclesCompleted(stats.getCyclesCompleted())
                    .workPercentage(stats.getWorkPercentage())
                    .currentPhase(stats.getCurrentPhase())
                    .events(stats.getEvents())
                    .build();

            Map<String, Object> result = new HashMap<>();
            result.put("stats", response);

            return ResponseEntity.ok(result);

        } catch (Exception e) {
            log.error("Failed to get timer stats", e);
            Map<String, Object> error = new HashMap<>();
            error.put("error", e.getMessage());
            return ResponseEntity.badRequest().body(error);
        }
    }
}
